### DISCUSSION TITLE
How does pipx manage environments for scripts with inline script metadata?

### DISCUSSION THREAD

**Discussion by jwodder** [pypa/pipx#1387] [Q&A] [answered]
Doing `pipx run somefile.py` where `somefile.py` specifies its dependencies via [inline script metadata](https://packaging.python.org/en/latest/specifications/inline-script-metadata/) causes `pipx` to create a virtualenv with those dependencies in order to run the script.  Running `pipx run somefile.py` will reuse the virtualenv most of the time, but sometimes `pipx` creates a new virtualenv despite the script's dependencies being unchanged; presumably, the old virtualenv was cleaned up automatically at some point.

How exactly does `pipx` decide when & which inline script venvs to purge?  Is there a maximum number of such venvs that can be present at any time?  Are venvs deleted if they were created or last used more than a certain amount of time ago?  Is there a way (or could a way be added) to configure the venv-purging parameters?

In my case, I have a number of small pipx-run scripts (mostly in `~/bin`) that I have no interest in making into packages.  These scripts are going to stay around for a while and will rarely have their dependencies modified, and it's a little annoying when a command that should complete in a fraction of a second instead takes several seconds due to pipx needing to create a virtualenv again.  One solution for this might be if `pipx install` could take a path to a local script; should I file that as a feature request?

**Comment by chrysle** [comment, marked as answer]
The cache duration for `pipx run` venvs is exactly 14 days, currently. For your case, would introducing an option to keep created virtual environments work? Probably we need some more fine-grained cache control, too, like `pip` has.

> One solution for this might be if pipx install could take a path to a local script; should I file that as a feature request?

This would definitely require some larger restructuring changes and IMO contrast with `pipx`'s current policy that only standalone command line applications can be installed (not sure if a simple script is worth the effort), but feel free to.

**Reply by jwodder** [reply]
> For your case, would introducing an option to keep created virtual environments work?

Pretty much, yes.

---

Feature request filed: https://github.com/pypa/pipx/issues/1388
